### PR TITLE
Planewise erosion, dilation, and opening

### PR DESCRIPTION
**Pull request by karhohs**
Previously #2643 , plane-wise processing of volumetric images was added to the Closing module. This same functionality has been added to the related morphological operations of erosion, dilation, and opening found in the modules of the same names.

### CODE REVIEW

**Comment by 0x00b1** [comment]
Awesome! I’ll review. First, you need tests. Second, what’s the motivation for planar processing?

**Comment by karhohs** [comment]
The justifications include those in #2643, but I'll reiterate them here. The primary concern is that current 3D structuring element doesn't respect the voxel spacing. For microscopy images this is problematic, because the z-spacing is typically much larger than the pixel size. Therefore, neighboring pixels in x,y are much more related to each other than pixels in the plane above and below. Treating them equally can lead to undesirable results.

Another argument is that since closing, erosion, dilation, and opening are all related to each other the modules should behave similarly for completeness and predictability within CellProfiler.

WRT to tests, will you please point to me to an existing test within CellProfiler that tests a 3d image that I can use as a template for new tests?

File: cellprofiler/modules/dilation.py [hunk 1]
```diff
@@ -42,10 +42,29 @@ def visible_settings(self):
...
+def planewise_morphology_dilation(x_data, structuring_element):
+    y_data = numpy.zeros_like(x_data)
+
+    for index, plane in enumerate(x_data):
+        if x_data.dtype == numpy.bool:
+            y_data[index] = skimage.morphology.binary_dilation(plane, structuring_element)
+        else:
+            y_data[index] = skimage.morphology.dilation(plane, structuring_element)
+
+    return y_data
```
Comment (mcquin) [inline cellprofiler/modules/dilation.py]: Newline required at end of file.

File: cellprofiler/modules/erosion.py [hunk 2]
```diff
@@ -42,10 +42,29 @@ def visible_settings(self):
...
+def planewise_morphology_erosion(x_data, structuring_element):
+    y_data = numpy.zeros_like(x_data)
+
+    for index, plane in enumerate(x_data):
+        if x_data.dtype == numpy.bool:
+            y_data[index] = skimage.morphology.binary_erosion(plane, structuring_element)
+        else:
+            y_data[index] = skimage.morphology.erosion(plane, structuring_element)
+
+    return y_data
```
Comment (mcquin) [inline cellprofiler/modules/erosion.py]: Newline required at end of file.

File: cellprofiler/modules/opening.py [hunk 3]
```diff
@@ -42,10 +42,28 @@ def visible_settings(self):
...
+def planewise_morphology_opening(x_data, structuring_element):
+    y_data = numpy.zeros_like(x_data)
+
+    for index, plane in enumerate(x_data):
+        if x_data.dtype == numpy.bool:
+            y_data[index] = skimage.morphology.binary_opening(plane, structuring_element)
+        else:
+            y_data[index] = skimage.morphology.opening(plane, structuring_element)
+
+    return y_data
```
Comment (mcquin) [inline cellprofiler/modules/opening.py]: Newline required at end of file.

**Comment by 0x00b1** [comment]
Interesting! I appreciate the explanation!

And I would look at the the tests for dilation, erosion, and opening. They include 3D test fixtures.

File: tests/modules/test_closing.py [hunk 4]
```diff
@@ -11,16 +12,41 @@ def test_run(image, module, image_set, workspace):
...
+        actual = image_set.get_image("closing")
+
+        desired = skimage.morphology.closing(image.pixel_data, selem)
+
+        numpy.testing.assert_array_equal(actual.pixel_data, desired)
+
+        # test planewise
+        selem = skimage.morphology.disk(1)
+
+        module.run(workspace)
```
Comment (mcquin) [inline tests/modules/test_closing.py]: Before `module.run(workspace)` you'll need to specify the new structuring element:

```python
module.structuring_element.shape = "disk"

module.run(workspace)
```

**Pull request merged**